src/com/google/javascript/jscomp/RewriteDynamicImports.java

```diff
@@ -0,0 +1,312 @@
+    if (n.getToken() != Token.DYNAMIC_IMPORT) {
+      return;
+    }
+
+    // If the module specifier is a string, attempt to resolve the module
+    final ModuleMap moduleMap = compiler.getModuleMap();
+    final Node importSpecifier = n.getFirstChild();
+    if (importSpecifier.isString() && moduleMap != null) {
+      final ModulePath targetPath = compiler.getModuleLoader().resolve(importSpecifier.getString());
+      final Module module = compiler.getModuleMap().getModule(targetPath);
+      final String targetModuleVarName =
+          (module == null)
+              ? null
+              : GlobalizedModuleName.create(module.metadata(), null, null).aliasName().join();
+      final Var targetModuleNS =
+          (targetModuleVarName == null) ? null : t.getScope().getVar(targetModuleVarName);
+
+      if (targetModuleNS != null) {
+        final JSModule targetModule = targetModuleNS.getInput().getModule();
+        // If the target module is bundled into the same output chunk, replace the import statement
+        // with a promise that resolves to the module namespace.
+        // No further rewriting occurs for this case.
+        if (t.getModule() == targetModule) {
+          replaceDynamicImportWithPromise(t, n, targetModuleNS);
+          return;
+        } else {
+          // The target output chunk is recognized and different from the current chunk.
+          // Retarget the import specifier path to the output chunk path and rewrite
+          // the import to reference the rewritten global module namespace variable.
+          retargetImportSpecifier(t, n, targetModule);
+          if (NodeUtil.isExpressionResultUsed(n)) {
+            addChainedThen(t, n, targetModuleNS);
+          }
+        }
+      }
+    }
+    if (this.alias != null) {
+      aliasDynamicImport(t, n);
+    }
+  }
+
+  /**
+   * Replace a dynamic import expression with a promise resolving to the rewritten module namespace.
+   *
+   * <p>Before
+   *
+   * <pre>
+   *   import('./foo.js');
+   * </pre>
+   *
+   * After
+   *
+   * <pre>
+   *   Promise.resolve(module$foo);
+   * </pre>
+   */
+  private void replaceDynamicImportWithPromise(
+      NodeTraversal t, Node dynamicImport, Var targetModuleNS) {
+    final Scope scope = t.getScope();
+    final JSTypeRegistry registry = compiler.getTypeRegistry();
+    final Node promiseDotResolve =
+        astFactory.createQName(scope.getGlobalScope(), "Promise.resolve");
+    final Node promiseResolveCall = astFactory.createCall(promiseDotResolve);
+    final boolean isExpressionUsed = NodeUtil.isExpressionResultUsed(dynamicImport);
+    if (isExpressionUsed) {
+      JSType moduleNamespaceType;
+      if (dynamicImport.getJSType() != null && dynamicImport.getJSType().isTemplatizedType()) {
+        TemplatizedType templatizedType = (TemplatizedType) dynamicImport.getJSType();
+        moduleNamespaceType = templatizedType.getTemplateTypes().asList().get(0);
```

Comment:
Could you drop the ".asList()" here? It shouldn't be necessary.

Reply:
Done